src/main/java/ca/nines/ise/writer/XMLWriter.java

```
@@ -16,179 +16,1303 @@
+	 * 
+	 * @param node
+	 *            node to copy attributes from
+	 * @param e
+	 *            destination Element to which attributes are added
+	 * @param changes
+	 *            HashMap of attributes whose values are to be changed
+	 * @param ignore
+	 *            Attribute names to be ignored
+	 * @param add
+	 *            extra Attributes to be added to @e
+	 * @return Element (@e)
+	 */
+	private Element set_attributes(TagNode node, Element e,
+			HashMap<String, String> changes, String[] ignore, Attribute[] add) {
+		for (String name : node.getAttributeNames()) {
+			if (ignore != null && Arrays.asList(ignore).contains(name))
+				continue;
+			if (changes != null && changes.containsKey(name))
+				e.addAttribute(new Attribute(changes.get(name), node
+						.getAttribute(name).replace(',', ' ')));
+			else
+				e.addAttribute(new Attribute(name, node.getAttribute(name)
+						.replace(',', ' ')));
+		}
+		if (add != null)
+			for (Attribute attr : add)
+				e.addAttribute(attr);
+		return e;
+	}
+
+	/**
+	 * removes bounding white space (strings which are entirely whitespace)
+	 * 
+	 * @param data
+	 * @return " " if boundary, @data otherwise
+	 */
+	private static String normalizeBoundary(String data) {
+		String temp = normalizeSpace(data);
+		if (temp.equals(" ") || temp.equals(""))
+			return temp;
+		return data;
+	}
+
+	private static String normalizeSpace(String data) {
+		data = data.replace('\t', ' ');
+		data = data.replace('\n', ' ');
+		data = data.replace('\r', ' ');
+		data = data.trim();
+
+		StringBuffer result = new StringBuffer();
+		for (int i = 0; i < data.length(); i++) {
+			if (i == 0 || data.charAt(i - 1) != ' ' || data.charAt(i) != ' ') {
+				result.append(data.charAt(i));
+			}
+		}
+
+		return result.toString();
+	}
+
+	@Override
+	public void render(DOM dom, Annotation ann)
+			throws TransformerConfigurationException, TransformerException,
+			IOException, Exception {
+		System.err.println("invalid renderer");
```

Yes, it should throw an exception like SGMLWriter does.